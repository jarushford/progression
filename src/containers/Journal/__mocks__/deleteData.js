export const deleteDataThunk = jest.fn()