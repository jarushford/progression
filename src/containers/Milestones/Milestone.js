import React from 'react'

export default function Milestone({ milestone_date, caption }) {
  return (
    <article className="milestone">
      <div>
        <h3>{milestone_date}</h3>
        <i className="fas fa-times"></i>
      </div>
      <p>{caption}</p>
    </article>
  )
}