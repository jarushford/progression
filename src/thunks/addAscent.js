import { fetchAscentsThunk } from './fetchAscents'
import { clearAscents, setError } from '../actions'

export const addAscentThunk = (ascent) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/ascents/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(ascent),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add ascent')
      }
      await dispatch(clearAscents())
      dispatch(fetchAscentsThunk(ascent.user_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}