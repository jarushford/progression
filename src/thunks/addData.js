import { clearThunkHelper } from '../utils/thunkHelpers'
import { fetchDataThunk } from './fetchData'
import { setError } from '../actions'

export const addDataThunk = (data, type) => {
  return async (dispatch) => {
    try {
      const url = `http://localhost:3000/api/progressionusers/${type}s/new`
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(data),
        headers: {
          "Content-Type": "application/json"
        }
      })

      if (!response.ok) {
        throw Error(`Could not add ${type}`)
      }

      await dispatch(clearThunkHelper(type)())
      type === 'journal' || type === 'milestone'
        ? dispatch(fetchDataThunk(data.user_id, data.project_id, type))
        : dispatch(fetchDataThunk(data.user_id, '', type))

      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}