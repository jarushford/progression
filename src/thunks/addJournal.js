import { fetchJournalThunk } from './fetchJournal'
import { clearJournal, setError } from '../actions'

export const addJournalThunk = (entry) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/journal/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(entry),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add entry')
      }
      await dispatch(clearJournal())
      dispatch(fetchJournalThunk(entry.user_id, entry.project_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}