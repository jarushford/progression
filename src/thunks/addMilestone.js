import { fetchMilestonesThunk } from './fetchMilestones'
import { clearMilestones, setError } from '../actions'

export const addMilestoneThunk = (milestone) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/milestones/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(milestone),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add project')
      }
      await dispatch(clearMilestones())
      dispatch(fetchMilestonesThunk(milestone.user_id, milestone.project_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}