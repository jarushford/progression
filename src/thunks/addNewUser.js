import { setError } from '../actions'
import { loginUserThunk } from './loginUser'

export const addNewUserThunk = (user) => {
  return async (dispatch) => {
    try {
      const url = `https://progression-backend.herokuapp.com/api/progressionusers/new`
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(user),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw new Error('User already exists, try another email!')
      }
      await dispatch(loginUserThunk(user))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}