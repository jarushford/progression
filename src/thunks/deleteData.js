import { clearThunkHelper } from '../utils/thunkHelpers'
import { fetchDataThunk } from './fetchData'
import { setError } from '../actions'

export const deleteDataThunk = (item_id, user_id, project_id, type) => {
  return async (dispatch) => {
    try {
      const data = {item_id, user_id, project_id}
      let url
      type === 'journal' || type === 'milestone'
        ? url = `http://localhost:3000/api/progressionusers/${user_id}/projects/${project_id}/${type}s/${item_id}`
        : url = `http://localhost:3000/api/progressionusers/${user_id}/${type}s/${item_id}`

      const response = await fetch(url, {
        method: 'DELETE',
        body: JSON.stringify(data),
        headers: {
          "Content-Type": "application/json"
        }
      })

      if (!response.ok) {
        throw Error(`Could not remove ${type}`)
      }

      await dispatch(clearThunkHelper(type)())

      type === 'journal' || type === 'milestone'
        ? dispatch(fetchDataThunk(data.user_id, data.project_id, type))
        : dispatch(fetchDataThunk(data.user_id, '', type))

    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}