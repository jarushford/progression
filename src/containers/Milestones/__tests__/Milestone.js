import React from 'react'
import { shallow } from 'enzyme'
import { deleteDataThunk } from '../../../thunks/deleteData'
import { mapDispatchToProps, Milestone } from '../Milestone'

jest.mock('../../../thunks/deleteData')

describe('Milestone', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<Milestone />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should delete an entry on click', () => {
    const mockDelete = jest.fn()
    const wrapper = shallow(<Milestone deleteMilestone={mockDelete} id={1} project_id={2} user_id={4} />)

    wrapper.find('.fa-times').simulate('click')

    expect(mockDelete).toBeCalledWith(1, 4, 2, 'milestone')
  })

  it('should return a props object with a method deleteMilestone', () => {
    const mockDispatch = jest.fn()
    const expected = deleteDataThunk(1, 2, 3, 'milestone')

    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.deleteMilestone(1, 2, 3, 'milestone')

    expect(mockDispatch).toBeCalledWith(expected)
  })
})