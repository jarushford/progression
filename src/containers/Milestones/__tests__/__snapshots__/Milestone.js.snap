// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Milestone should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Milestone
    caption="Test caption"
    deleteMilestone={[MockFunction]}
    id={1}
    milestone_date="1/1/12"
    project_id={2}
    user_id={4}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <h3>
            1/1/12
          </h3>
          <i
            className="fas fa-times"
            onClick={[Function]}
          />
        </div>,
        <p
          className="milestone-cap"
        >
          Test caption
        </p>,
      ],
      "className": "milestone",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3>
              1/1/12
            </h3>,
            <i
              className="fas fa-times"
              onClick={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "1/1/12",
            },
            "ref": null,
            "rendered": "1/1/12",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fas fa-times",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Test caption",
          "className": "milestone-cap",
        },
        "ref": null,
        "rendered": "Test caption",
        "type": "p",
      },
    ],
    "type": "article",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <h3>
              1/1/12
            </h3>
            <i
              className="fas fa-times"
              onClick={[Function]}
            />
          </div>,
          <p
            className="milestone-cap"
          >
            Test caption
          </p>,
        ],
        "className": "milestone",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3>
                1/1/12
              </h3>,
              <i
                className="fas fa-times"
                onClick={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "1/1/12",
              },
              "ref": null,
              "rendered": "1/1/12",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fas fa-times",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Test caption",
            "className": "milestone-cap",
          },
          "ref": null,
          "rendered": "Test caption",
          "type": "p",
        },
      ],
      "type": "article",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
